R/R/diagnostics.R

```diff
@@ -67,49 +67,62 @@ generate_cutoffs <- function(df, horizon, initial, period) {
 #'  horizon. If not provided, 0.5 * horizon is used.
 #' @param initial Integer size of the first training period. If not provided,
 #'  3 * horizon is used. Same units as horizon.
+#' @param cutoffs Vector of cutoff dates to be used during
+#'  cross-validtation. If not provided works beginning from (end - horizon),
+#'  works backwards making cutoffs with a spacing of period until initial is
+#'  reached.
 #'
 #' @return A dataframe with the forecast, actual value, and cutoff date.
 #'
 #' @export
 cross_validation <- function(
-    model, horizon, units, period = NULL, initial = NULL) {
+    model, horizon, units, period = NULL, initial = NULL, cutoffs=NULL) {
   df <- model$history
   horizon.dt <- as.difftime(horizon, units = units)
-  # Set period
-  if (is.null(period)) {
-    period <- 0.5 * horizon
+
+  predict_columns <- c('ds', 'yhat')
+  if (model$uncertainty.samples){
+    predict_columns <- append(predict_columns, c('yhat_lower', 'yhat_upper'))
   }
-  period.dt <- as.difftime(period, units = units)
   # Identify largest seasonality period
   period.max <- 0
   for (s in model$seasonalities) {
     period.max <- max(period.max, s$period)
   }
   seasonality.dt <- as.difftime(period.max, units = 'days')
-  # Set initial
-  if (is.null(initial)) {
-    initial.dt <- max(
-      as.difftime(3 * horizon, units = units),
-      seasonality.dt
-    )
-  } else {
-    initial.dt <- as.difftime(initial, units = units)
-    if (initial.dt < seasonality.dt) {
-      warning(paste0('Seasonality has period of ', period.max, ' days which ',
-        'is larger than initial window. Consider increasing initial.'))
+
+  if (is.null(cutoffs)){
+
+    # Set period
+    if (is.null(period)) {
+      period <- 0.5 * horizon
     }
-  }
-  
-  predict_columns <- c('ds', 'yhat')
-  if (model$uncertainty.samples){
-    predict_columns <- append(predict_columns, c('yhat_lower', 'yhat_upper'))
+    period.dt <- as.difftime(period, units = units)
+    # Set initial
+    if (is.null(initial)) {
+      initial.dt <- max(
+        as.difftime(3 * horizon, units = units),
+        seasonality.dt
+      )
+    }else {
+      initial.dt <- as.difftime(initial, units = units)
+    }
+    cutoffs <- generate_cutoffs(df, horizon.dt, initial.dt, period.dt)
+  }else{
+    initial.dt <- cutoffs[1] - min(df$ds)
   }
 
-  cutoffs <- generate_cutoffs(df, horizon.dt, initial.dt, period.dt)
+  # Check if the initial window  (that is, the amount of time between the
+  # start of the history and the first cutoff) is less than the
+  # maximum seasonality period
+  if (initial.dt < seasonality.dt) {
+    warning(paste0('Seasonality has period of ', period.max, ' days which ',
+      'is larger than initial window. Consider increasing initial.'))
+  }
 
   predicts <- data.frame()
   for (i in 1:length(cutoffs)) {
-    cutoff <- cutoffs[i]
+    cutoff <- as.POSIXct(cutoffs[i])
```

Comment:
I was doing some testing of this to figure out what the warnings were about and noticed this small but important thing: This doesn't handle time-zones well. If you do `as.POSIXct(as.Date('2010-01-01'))` what you will get depends on your timezone. I get `[1] "2009-12-31 16:00:00 PST"`, which is not the date that I put in. In the rest of Prophet we strip out time zones and replace them all with GMT (which is what as.Date produces). That's what we'll want to use here. Instead of using `as.POSIXct`, we should here do
```
cutoffs <- set_date(ds=cutoffs)
```
where `set_date` is coming from here: https://github.com/facebook/prophet/blob/cb4f1dcc59e61bb292d6cefab8e61b2f77585088/R/R/prophet.R#L250

That will make sure that these dates are handled in the same way as those in the history.

Reply:
(this would go in L112 above, at the start of the `else` for if cutoffs were specified)